Add a global option --retry-lock

jooola opened the pull request:
Attempt to resurrect https://github.com/restic/restic/pull/2214
 

What does this PR change? What problem does it solve?
-----------------------------------------------------

If the option is passed, restic will wait the specified duration of time and retry locking the repo every 10 seconds (or more often if the total timeout is relatively small).

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Fixes restic#719

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

MichaelEischer commented on cmd/restic/lock.go:
```diff
@@ -43,7 +43,38 @@ func lockRepository(ctx context.Context, repo restic.Repository, exclusive bool)
+					}
+					retryMessagePrinted = true
+				}
+				time.Sleep(retrySleep)
```
The loop should also exit if the context is canceled.

MichaelEischer replied:
The `time.Sleep` statement will block until the sleep duration has ended, even when the context was canceled in the meantime. In particular, if the sleep duration is longer than 5 seconds, that could lead to a significant delay (although ctrl-c leads to a call to exit at some point, which should bypass this during normal usage)

jooola replied:
I had the same thinking that sleep should be blocking. I wrote a test to check we exit on cancel, but I might have made a mistake in it, as it was exiting on time.

MichaelEischer commented on cmd/restic/lock.go:
```diff
@@ -21,17 +21,17 @@ var globalLocks struct {
 // lockRepository wraps the ctx such that it is cancelled when the repository is unlocked
 // cancelling the original context also stops the lock refresh
-func lockRepository(ctx context.Context, repo restic.Repository, exclusive bool) (*restic.Lock, context.Context, error) {
+func lockRepository(ctx context.Context, repo restic.Repository, exclusive bool, gopts GlobalOptions) (*restic.Lock, context.Context, error) {
```
when looking at the tests, I actually think that it's better to just manually pass the two necessary attributes of `GlobalOptions` into `lockRepository`. That is please replace `gopts GlobalOptions` with `retryLock time.Duration, json bool`.

MichaelEischer commented on cmd/restic/lock.go:
```diff
@@ -43,7 +43,38 @@ func lockRepository(ctx context.Context, repo restic.Repository, exclusive bool)
+	var lock *restic.Lock
+	var err error
+
+	retrySleep := 5 * time.Second
```
Retrying every five seconds, appears to be a bit frequent, especially if a longer RetryLock duration is used. What about increasing the duration depending on the RetryLock duration?

jooola replied:
I have to admit I have no educated guess for those values.

What about we start with 5s and bump the sleep duration at every loop until we reach the max of 5 minutes ?
Some sort of exponential back off ?

Is 5m too long ? Should we reduce the max value to 1 minute ?

MichaelEischer replied:
A maximum of 1 minute is probably better. Waiting for 5 minutes requires quite a bit of patience. 

What might be useful is to bound the value of `retrySleepNext` such that it does not exceed the retryLock limit (but only if the duration doesn't be come too short).

jooola replied:
> What might be useful is to bound the value of retrySleepNext such that it does not exceed the retryLock limit (but only if the duration doesn't be come too short).

I am not sure to understand this one. From what I understand we don't need to reduce `retrySleepNext` below `retryLock` because the `retryTimeout` (based on retryLock) will exit before the sleeping period.

We do make sure `retrySleep` is smaller than `retryLock` before entering the loop.

MichaelEischer commented on cmd/restic/lock_test.go:
```diff
@@ -168,3 +170,47 @@ func TestLockSuccessfulRefresh(t *testing.T) {
+	elock, _, err := lockRepoExclusive(context.TODO(), repo, env.gopts)
+	test.OK(t, err)
+
+	wait := time.Second
```
can we reduce that duration further? Our tests already take longer than I'd like.

MichaelEischer commented on cmd/restic/lock_test.go:
```diff
@@ -168,3 +170,47 @@ func TestLockSuccessfulRefresh(t *testing.T) {
+	})
+
+	env.gopts.RetryLock = wait
+	lock, _, err := lockRepo(context.TODO(), repo, env.gopts)
```
That will still sleep for 5 seconds, won't it?

jooola replied:
Yes indeed, I reduced the tests duration for the retry lock feature to 0.6s.

```
=== RUN   TestLockWaitTimeout
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    lock.go:102: setting lock timeout to 0s
    integration_test.go:56: repository initialized at /tmp/restic-test-3574893611/repo
--- PASS: TestLockWaitTimeout (0.25s)
=== RUN   TestLockWaitCancel
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    lock.go:102: setting lock timeout to 0s
    integration_test.go:56: repository initialized at /tmp/restic-test-1146899562/repo
--- PASS: TestLockWaitCancel (0.18s)
=== RUN   TestLockWaitSuccess
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    testing.go:32: using low-security KDF parameters for test
    config.go:74: disabling check of the chunker polynomial
    lock.go:102: setting lock timeout to 0s
    integration_test.go:56: repository initialized at /tmp/restic-test-883179798/repo
--- PASS: TestLockWaitSuccess (0.19s)
PASS
ok  	github.com/restic/restic/cmd/restic	0.628s
```

MichaelEischer reviewed (changes requested):
The PR needs a few more tweaks, but it's already in a good shape.

jooola committed:
Add a global option --retry-lock

Fixes restic#719

If the option is passed, restic will wait the specified duration of time
and retry locking the repo every 10 seconds (or more often if the total
timeout is relatively small).

- Play nice with json output
- Reduce wait time in lock tests
- Rework timeout last attempt
- Reduce test wait time to 0.1s
- Use exponential back off for the retry lock
- Don't pass gopts to lockRepo functions
- Use global variable for retry sleep setup
- Exit retry lock on cancel
- Better wording for flag help
- Reorder debug statement
- Refactor tests
- Lower max sleep time to 1m
- Test that we cancel/timeout in time
- Use non blocking sleep function
- Refactor into minDuration func

Co-authored-by: Julian Brost <julian@0x4a42.net>

MichaelEischer reviewed (approved):
LGTM. Please squash the commits.

I wonder whether the timing checks in the tests might be too strict or not. But we can fix that if it turns out to be a problem.

jooola commented:
- Rebased onto the master branch
- Squashed the PR history into the first commit

MichaelEischer committed:
lock: increase test timeout tolerances to avoid test failures

MichaelEischer commented:
Hmm, it looks like the durations checks are too strict:
https://github.com/restic/restic/actions/runs/4638924198/jobs/8211940279?pr=4107
https://github.com/restic/restic/actions/runs/4638924198/jobs/8211506876?pr=4107

I've broadened them quite a bit, but I fear these might become another `TestLockSuccessfulRefresh` which fails far too often.

MichaelEischer committed:
Merge branch 'master' into feature-wait-for-unlock
